---
title: "Pipeline GSE145408"
output: html_notebook
---

```{r eval=false}

library(RColorBrewer)
library(readxl)
library(UpSetR)
library(gplots)
library(ggupset)
library(ggrepel)
library(ggplot2)
library(pheatmap)
library(GEOquery)
library(readr)
#library(FactoMineR)
library(forcats)
library(heatmap3)
library(Biobase)
library(tools)
library(AnnotationDbi)
library(ade4)
library(ALL)
library(gcrma)
library(limma)
library(oligo)
library(affy)
library(genefilter)
library(RVAideMemoire)
library(enrichplot)
library(annotate)
#library(Affyhgu133aExpr)
library(hgu133a.db)
library(pd.huex.1.0.st.v2)
library(pd.ht.hg.u133.plus.pm)

library(dbplyr)
library(affycoretools)
library(dplyr)

# Fonction pour enregistrer les pheatmap (https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file)
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
    stopifnot(!missing(x))
    stopifnot(!missing(filename))
    pdf(filename, width=width, height=height)
    grid::grid.newpage()
    grid::grid.draw(x$gtable)
    dev.off()
}

```


```{r eval=false}
#Data import#####
wd_a <- "/Users/olivier/Desktop/Projet/Test/"
knitr::opts_chunk$set(root.dir = wd_a)
 # Accessing Raw Data from GEO:
my_id = "GSE145408"
filePaths = getGEOSuppFiles(my_id)
```


```{r eval=false}
### Import data ----

#Lecture du fichier Target
file_pheno_xlsx <-paste0(my_id,"_Target.xlsx")
file_pheno <- paste0(my_id,"_Target.txt")

data <- read_excel(paste0(wd_a,my_id,"/",file_pheno_xlsx))

write.table(data, file = file_pheno, sep = "\t", row.names = TRUE)


pd <- read.AnnotatedDataFrame(file = file_pheno, row.names = 1, header = TRUE, as.is = TRUE, sep = "\t", col.names = c("GSM", "Group", "Patient", "Time", "Group_short", "Description", "File"))

# Iterate through each column of pd
for (col in colnames(pd)) {
  # Remove double quotes from the column
  pd[[col]] <- gsub("\"", "", pd[[col]])
}

```

```{r}
#Changement de wd (besoin d'extraire le fichier RAW)
wd <- paste0(wd_a,my_id,"/",my_id,"_RAW")
df1 <- getGSEDataTables #??


knitr::opts_chunk$set(root.dir = wd)
write.table(matrix(dir()),file="a-list-files.txt", sep="\t")
```

```{r}
#Reads CEL files


# Lectures des fichiers .CEL
rawAffyData <- read.celfiles(filenames = pData(pd)$File, phenoData = pd)

# Vérification du bon import et première visualisation des données brutes
exprs(rawAffyData)[1:4,1:3]
boxplot(rawAffyData, which=c("all"))
hist(rawAffyData)
hist(rawAffyData, which=c("all"))
#pmSeq <- pmSequence(rawAffyData)
#pmSeq[1:5]
colnames(rawAffyData)


# Normalizes data ####
#Summarised per gene:
eset <- oligo::rma(rawAffyData) #Normalization entre les puces
#boxplot(eset)
#change directory: ####


# annotate: #### Selon la puce, changer le package d'annotation
#library("pd.ht.hg.u133.plus.db")
# eset <-annotateEset(eset,pd.ht.hg.u133.plus.db)
geoFile<-getGEO(my_id)
fData(eset) <- fData(geoFile$GSE145408_series_matrix.txt.gz)
colnames(fData(eset))

# Vérification de la bonne annotation
geneName = "GAPDH"
which(fData(eset)$"Gene Symbol" == geneName )
#head(fData(eset))


# filtration:
eset_medians <- rowMedians(Biobase::exprs(eset))
hist_res <- hist(eset_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities")
median(eset_medians)
man_threshold =median(eset_medians)
abline(v = median(eset_medians), col = "grey", lwd = 2)
abline(v = man_threshold, col = "coral4", lwd = 2)

length(which((rowMedians(exprs(eset)) >= man_threshold) == "TRUE"))

#Filtration selon la médiane d'expression (enleve aussi les NA)
data_filtr  <- eset[rowMedians(exprs(eset)) >= man_threshold,]
colnames(data_filtr) <- pData(data_filtr)$GSM
dim(data_filtr)
dim(eset)

#Rassemble tout dans un dataset
total <- cbind(fData(data_filtr), exprs(data_filtr))
dim(total)

#Export des données d'expression
#write.table(total, file=paste0("Données d'expression du ",gse_id), sep="\t")

# Suppression des genes mal annotés -> ne fonctionne pas pour l'instant
if (length(which(is.na(fData(data_filtr)$"Gene Symbol"))) != 0 ){
index <- which(is.na(fData(data_filtr)$"Gene Symbol"))
sansNA <- data_filtr[-index, ]
length(which(is.na(fData(sansNA)$"Gene Symbol") ==T ))
dim(sansNA)
dim(data_filtr)
data_filtr <- sansNA
sansNA_medians <- rowMedians(Biobase::exprs(sansNA))
hist_res <- hist(sansNA_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                border = "antiquewhite4",
                 xlab = "Median intensities")
}
dim(data_filtr)


#data_filter <- nsFilter(eset = data_filtr, reqire.entrez=TRUE, remove.dupEntrez=TRUE)


##### END import #####
```

## Test
```{r}
# Load required libraries
library(dplyr)

expression_data <- data_filtr@assayData[["exprs"]]
probe_ids <- rownames(expression_data)
probe_to_symbol <- data_filtr@featureData@data %>%
  select(ID, `Gene Symbol`)

probe_to_symbol_map <- setNames(probe_to_symbol$`Gene Symbol`, probe_to_symbol$ID)
gene_symbols <- probe_to_symbol_map[probe_ids]
expression_data <- cbind(expression_data, `Gene Symbol` = gene_symbols)
expression_data <- as.data.frame(expression_data)

expression_data$`Gene Symbol` <- as.character(expression_data$`Gene Symbol`)

## Suppression des genes dupliqués (plusieurs probes pour le meme gene)
duplicated_rows <- duplicated(expression_data$`Gene Symbol`)
expression_data_unique <- expression_data[!duplicated_rows, ]


## Si on veut calculer la moyenne (long)
# expression_data_mean <- expression_data %>%
#   group_by(`Gene Symbol`) %>%
#   summarize(across(starts_with("GSM"), mean, na.rm = TRUE))

dim(data_filtr)
dim(expression_data_unique)
#dim(expression_data_mean)

indexDup <- rownames(expression_data_unique)
sansDup <- data_filtr[indexDup,]
data_filtr <- sansDup

```




## Analyse
```{r}
summary(exprs(data_filtr))
boxplot(exprs(data_filtr),outline=FALSE)
```
```{r}

```

# Inspection des variables
```{r}
sampleInfo <- pData(data_filtr)
colnames(sampleInfo)
rownames(sampleInfo) <- sampleInfo$GSM
## Utilisation des infos contenues dans source_name_ch1 et characteristics_ch1.1 
sampleInfo <- dplyr::select(sampleInfo, Patient,Time,Group_short,Group,Description,GSM) 

## Si jamais on a besoin de renommer nos variables (plus lisible)
#sampleInfo <- dplyr::rename(sampleInfo,group = source_name_ch1, patient=characteristics_ch1)
View(sampleInfo)
```


# Heatmap entre GSM
```{r}
#Calcul de la matrice de correlation d'expression des gènes pour chaque GSM (étude de correlation entre les samples)
corMatrix <- cor(exprs(data_filtr),use="c")

## Vérifier que les colonnes de sampleInfo et corMatrix soient identiques
rownames(sampleInfo)
colnames(corMatrix)

## Sinon, on peut forcer avec le code suivant
rownames(sampleInfo) <- colnames(corMatrix)


# Création de notre vecteur annotations pour les heatmaps
annotations = subset(sampleInfo, select = c(Group_short,Time))

# Récupérer l'ordre des lignes en fonction de sampleInfo$GSM
order_rows <- order(sampleInfo$GSM)

# Réorganiser les lignes de la matrice de corrélation
corMatrix_ordered <- corMatrix[order_rows, order_rows]

# Créer la heatmap en utilisant pheatmap avec les lignes ordonnées #ne fonctionne pas pour l'instant
heatmapGSM_nofiltr <- pheatmap(corMatrix_ordered, 
         annotation_col = annotations,
         main = paste0(my_id,' - Heatmap - GSM with outliers'))


save_pheatmap_pdf(heatmapGSM_nofiltr, paste0(my_id,'_heatmap-GSM-with-outliers.pdf'))
```

# Export

```{r eval=false}
#full_output <- cbind(fData(data_filtr),exprs(data_filtr))
#write_csv(full_output, path="gse_full_output.csv")
```

# Définition de nos features (et modification des noms de features si besoin (selon version de puces))
```{r}
library(tidyverse)
#Import de features à partir des données fData
features <- fData(data_filtr)
colnames(features)

## Pour garder que certaines features
features <- dplyr::select(features,"Gene Symbol","Gene Title","ENTREZ_GENE_ID")

features <- features %>% 
  dplyr::rename(GENE_SYMBOL = `Gene Symbol`,
  GENE_TITLE = `Gene Title`)

#Affichage de features
View(features)

#Export des features
full_output <- cbind(features,exprs(data_filtr))
write_csv(full_output, path="gse_full_output.csv")
```



# ACP
```{r}
## Attention à bien avoir transposé la matrice d'expression sinon PCA en fonction des individus et non de l'expression des genes

#PCA classique
pca <- prcomp(t(exprs(data_filtr)))

PoV <- pca$sdev^2/sum(pca$sdev^2)
PoV <- PoV * 100
PoV <- substr(PoV,1,4)
PC1_t = paste0("PC1 (", PoV[1] , "%)")
PC2_t = paste0("PC2 (", PoV[2] ,"%)")
##Rajouter les informations des patients
cbind(sampleInfo, pca$x) %>% 
ggplot(aes(x = PC1, y=PC2, col=Group,label=GSM)) + 
  geom_point() + 
  geom_text_repel() +
  labs(x = PC1_t , y = PC2_t, color = "Treatment") +
  labs(title = paste0(my_id, ' - PCA - GSM') )
```

```{r}
#PCA en utilisant FactoMineR
#pca_bis <- PCA(X = t(exprs(data_filtr)), ncp = 5 , graph= FALSE)
```


```{r} 
## A MODIFIER
# 
# ##On va nous même faire le graphe pour montrer l'évolution de chaque sample avant changement (T0) et post changement de traitement (T1)
# # Plot PCA
# 
# # Création de notre dataframe pour l'affichage de l'ACP
# df_pca <- data.frame(
#   Patient = sampleInfo$GSM,
#   Time = sampleInfo$Group,
#   PC1 = pca_bis$ind$coord[, 1], #Coordonnées pour PC1
#   PC2 = pca_bis$ind$coord[, 2]  ##Coordonnées pour PC2
# )
# 
# 
# # Créer un dataframe pour les flèches reliant T0 et T1 pour chaque patient
# df_arrows <- df_pca %>%
#   mutate(PC1_end = lead(PC1),
#          PC2_end = lead(PC2)) %>%
#   filter(!is.na(PC1_end))  # Supprimer les lignes avec des valeurs manquantes
# 
# 
# #Titre des axes
# PC1_t = paste0("PC1 (", substr(pca_bis$eig[1,2],1,4) , "%)")
# PC2_t = paste0("PC2 (", substr(pca_bis$eig[2,2],1,4) ,"%)")
# 
# 
# # Plot PCA avec une seule flèche entre chaque paire de patients à T0 et T1
# pca_plot <- ggplot(
#   df_pca, aes(x = PC1, y = PC2, color = Time)) +
#   geom_point(size = 3) +
#   geom_segment(data = df_arrows[seq(1, nrow(df_arrows), by = 2), ], 
#                aes(xend = PC1_end, yend = PC2_end),
#                arrow = arrow(length = unit(0.1, "inches")), alpha = 0.5) +
#   geom_text(aes(label = Patient), vjust = -1) +
#   geom_hline(yintercept = 0, linetype = "dotted") +
#   geom_vline(xintercept = 0, linetype = "dotted") +
#   ggtitle("ACP avec l'évolution des patients entre T0 et T1") +
#   labs(x = PC1_t , y = PC2_t, color = "Time") +
#   theme_minimal()
# 
# 
# # Afficher le plot de l'ACP
# print(pca_plot)
```


## Differential Expression

# Création de notre matrice design et constrast
```{r}

sampleInfo <- sampleInfo %>%
  mutate(method = case_when(Group == "NW" ~ "NoWithdrawal",
                                Group == "WR" | Group == "WS" ~ "Withdrawal")) 


sampleInfo <- sampleInfo %>%
  mutate(method = fct_relevel(method, "NoWithdrawal", "Withdrawal")) 


##Création de la matrice de design 
# ici on compare no withdrawal (N=4) and Tac withdrawal (N=14) + 3months post withdrawal
design <- model.matrix(~0+ method, data = sampleInfo) 

# Afficher la matrice de design
print(design)

## Si on fait Retrait vs Non Retrait
comparison <- "DiffNW-WD"
contrasts <- makeContrasts( Comparaison= methodNoWithdrawal - methodWithdrawal, levels = design)
```

# lmFit
```{r}
fit <- lmFit(exprs(data_filtr), design)

#Vérifier qu'il n'y a pas de pb (sur les premiers coefficients)
head(fit$coefficients)

# On applique nos constrasts 
fit2 <- contrasts.fit(fit, contrasts)
fit2

# On utilise *empirical Bayes'* pour avoir les p-values de chaque gènes (première estimation?)
fit2 <- eBayes(fit2)

# Résultats avec topTable
topTable(fit2)

#decideTests donne cb de genes sont differentially expressed
decideTests(fit2)
table(decideTests(fit2))
```

# Visualisation des resultat de la differencial expression

```{r}
fit2$genes <- features
summary(topTable(fit2))

full_results <- topTable(fit2, number=Inf)
full_results <- tibble::rownames_to_column(full_results,"ID")
View(full_results)

# AJOUT DE LA CATEGORIE UP/DOWN OU NON SIGNIFICATIF
fc_cutoff <- log2(1.5)
p_cutoff <- 0.05


full_results <- full_results %>%
  dplyr::mutate(gene_type = case_when(logFC >= fc_cutoff & adj.P.Val <= 0.05 ~ "up",
                               logFC <= -fc_cutoff & adj.P.Val <= 0.05 ~ "down",
                               TRUE ~ "ns")) 

cols <- c("up" = "gold", "down" = "blue", "ns" = "darkgrey") 
#???
# full_results <- full_results %>%
#   dplyr::mutate(gene_type = fct_relevel(gene_type, "up", "down")) 
```

# Volcano plots
```{r}
full_results %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type)) + geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, " - Volcano plot"))
ggsave(paste0(my_id,'-Volcano-plot.pdf'))
```
## Affichage des 20 gènes ayant leur expression la plus modifiée

```{r}

topN <- 20 #Nombre de genes affichés sur le graphe
full_results %>% 
  dplyr::mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type,label=Label)) + 
  geom_point() +    
  geom_text_repel(col = "black", max.overlaps = 25)  + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, ' - Volcano plot - Top ',topN,' genes '))
ggsave(paste0(my_id,'-Volcano-Top20-plot.pdf'))


```
## Affichage des 20 gènes de la tolérance (Baron et al. 2015)
```{r}
my_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")

full_results %>% 
  dplyr::mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>%
  dplyr::mutate(Rank = rank(dplyr::desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = gene_type, label = Label)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey")) + 
  geom_text_repel(col = "black", max.overlaps = 5000) +
  labs(title="Volcano plot - 20 genes of tolerance (Baron et al. 2015)")
ggsave(paste0(my_id,'-Volcano-plot-20-genes-TOL.pdf'))


```

# Export avec write_csv (filtrer ceux significatifs)

```{r}
filter(full_results, adj.P.Val < 0.05, abs(logFC) > fc_cutoff ) %>%
  write_csv(path=paste0(my_id,"_filtered_significant_results.csv"))
```


### Autre visualisation (Heatmaps)


# Heatmap des 20 genes les plus significatifs
```{r}
## Top N genes
topN <- 20

top_ids_of_interest <- mutate(full_results, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(ID)

#On recupère le nom des top N genes
top_gene_names <- mutate(full_results, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(GENE_SYMBOL) 

## Recupere les expressions des genes d'interet
gene_matrix_top <- exprs(data_filtr)[top_ids_of_interest,]
write.csv(gene_matrix_top, file=paste0(my_id,"_filtered_results_topN_genes.csv"))

##Affichage de la heatmap des topN genes
heatmap_topN_genes <- pheatmap(gene_matrix_top,
     labels_row = top_gene_names,
     scale="row",
     annotation_col = annotations,
     main = paste0(my_id, ' - Heatmap - Top ',topN,' 20 genes'))

save_pheatmap_pdf(heatmap_topN_genes, paste0(my_id,'_heatmap_topN_genes.pdf'))

#PROBLEME MANQUE DEUX NOM DE GENES????
```


### Heatmap 20 genes de la tolérance (BARON)

```{r}
#20 GENES (BARON 2015)
Baron_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")

Baron_ids_of_interest <-  filter(full_results,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(ID)

Baron_gene_names <-  filter(full_results,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(GENE_SYMBOL)

Baron_gene_matrix <- exprs(data_filtr)[Baron_ids_of_interest,]
write.csv(x = Baron_gene_matrix, file = paste0(my_id,"_filtered_results_20_genes.csv"))

#Affichage
rownames(sampleInfo) <- colnames(Baron_gene_matrix)
heatmap_20_genes <- pheatmap(Baron_gene_matrix,
         labels_row = Baron_gene_names,
         annotation_col= annotations,
         scale="row",
         main = paste0(my_id, ' - Heatmap - 20 genes of tolerance (Baron et al. 2015)'))

save_pheatmap_pdf(heatmap_20_genes, paste0(my_id,'_heatmap_20_genes.pdf'))
```


```{r}
#6 GENES (DANGER 2017)
Danger_genes <- c("AKR1C3", "CD40","CTLA4","ID3", "MZB1", "TCL1A")

Danger_ids_of_interest <-  filter(full_results,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(ID)

Danger_gene_names <-  filter(full_results,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(GENE_SYMBOL)  


#Affichage
Danger_gene_matrix <- exprs(data_filtr)[Danger_ids_of_interest,]
write.csv(x = Danger_gene_matrix, file=paste0(my_id,"_filtered_results_6_genes.csv"))


heatmap_6_genes <- pheatmap(Danger_gene_matrix,
         labels_row = Danger_gene_names,
         annotation_col= annotations,
         main = paste0(my_id, ' - Heatmap - 6 genes of tolerance (Danger et al. 2017)'),
         scale="row")

save_pheatmap_pdf(heatmap_6_genes, paste0(my_id,'_heatmap_6_genes.pdf'))
```
## Analyse T0 vs T1 pour WR, WS et NW

# Création de notre matrice design et constrast
```{r}
sampleInfo <- sampleInfo %>%
  mutate(method = case_when(Group == "NW" ~ "NoWithdrawal",
                                Group == "WR" | Group == "WS" ~ "Withdrawal")) 

sampleInfo <- sampleInfo %>%
  mutate(method = fct_relevel(method, "NoWithdrawal", "Withdrawal")) 

index_WS <- which(data_filtr$Group == "WS")
index_WR <- which(data_filtr$Group == "WR")
index_NW <- which(data_filtr$Group == "NW")
data_WS <- data_filtr[,index_WS]
data_WR <- data_filtr[,index_WR]
data_NW <- data_filtr[,index_NW]
dim(data_WS)
dim(data_WR)
dim(data_NW)
dim(data_filtr)

##Création de la matrice de design 
# ici on compare T0 et T1 pour chacun des groupes
design_WS <- model.matrix(~0+ Time, data = sampleInfo[index_WS,])
design_WR <- model.matrix(~0+ Time, data = sampleInfo[index_WR,])
design_NW <- model.matrix(~0+ Time, data = sampleInfo[index_NW,])
# Afficher la matrice de design
print(design_WS)
print(design_WR)
print(design_NW)

```


## Etablissement de la différential expression

# lmFit
```{r}
fit_WS <- lmFit(exprs(data_WS), design_WS)
fit_WR <- lmFit(exprs(data_WR), design_WR)
fit_NW <- lmFit(exprs(data_NW), design_NW)


```


#Définition de nos contrasts pour la differencial expression

```{r}

## On compare T0 et T1 pour chacun des sous ensembles
contrasts_WS <- makeContrasts(EffetTimeWS = TimeT0 - TimeT1,
                           levels=colnames(design_WS))
contrasts_WR <- makeContrasts(EffetTimeWR = TimeT0 - TimeT1,
                           levels=colnames(design_WR))
contrasts_NW <- makeContrasts(EffetTimeNW = TimeT0 - TimeT1,
                           levels=colnames(design_NW))

# On applique nos constrasts 

fit2_WS <- contrasts.fit(fit_WS, contrasts_WS)
fit2_WR <- contrasts.fit(fit_WR, contrasts_WR)
fit2_NW <- contrasts.fit(fit_NW, contrasts_NW)


```

# On utilise *empirical Bayes'* pour avoir les p-values de chaque gènes (première estimation?)
```{r}
fit2_WS <- eBayes(fit2_WS)
fit2_WR <- eBayes(fit2_WR)
fit2_NW <- eBayes(fit2_NW)


# Résultats avec topTable
topTable(fit2_WS)
comparisonWS = "EffetTimeWS"
topTable(fit2_WS, coef=comparisonWS, adjust.method="BH")
#decideTests donne cb de genes sont differentially expressed
decideTests(fit2_WS)
table(decideTests(fit2_WS))

# Résultats avec topTable
topTable(fit2_WR)
comparisonWR = "EffetTimeWR"
topTable(fit2_WR, coef=comparisonWR, adjust.method="BH")
#decideTests donne cb de genes sont differentially expressed
decideTests(fit2_WR)
table(decideTests(fit2_WR))

# Résultats avec topTable
topTable(fit2_NW)
comparisonNW = "EffetTimeNW"
topTable(fit2_NW, coef=comparisonNW, adjust.method="BH")
#decideTests donne cb de genes sont differentially expressed
decideTests(fit2_NW)
table(decideTests(fit2_NW))
```


# Visualisation des resultat de la differencial expression

```{r}
fit2_WS$genes <- features
summary(topTable(fit2_WS))

fit2_WR$genes <- features
summary(topTable(fit2_WR))

fit2_NW$genes <- features
summary(topTable(fit2_NW))
```

```{r}
#full_results <- topTable(fit2, number=Inf)
full_results_WS <- topTable(fit2_WS, coef=comparisonWS, adjust.method="BH", number=Inf)
full_results_WR <- topTable(fit2_WR, coef=comparisonWR, adjust.method="BH", number=Inf)
full_results_NW <- topTable(fit2_NW, coef=comparisonNW, adjust.method="BH", number=Inf)

full_results_WS <- tibble::rownames_to_column(full_results_WS,"ID")
full_results_WR <- tibble::rownames_to_column(full_results_WR,"ID")
full_results_NW <- tibble::rownames_to_column(full_results_NW,"ID")

View(full_results_WS)
View(full_results_WR)
View(full_results_NW)
```



# Visualisation des resultat de la differencial expression

```{r}
# AJOUT DE LA CATEGORIE UP/DOWN OU NON SIGNIFICATIF
fc_cutoff <- log2(1.5)
p_cutoff <- 0.05

## WS - Time T0 vs T1
full_results_WS <- full_results_WS %>%
  mutate(gene_type = case_when(logFC >= fc_cutoff & adj.P.Val <= 0.05 ~ "up",
                               logFC <= -fc_cutoff & adj.P.Val <= 0.05 ~ "down",
                               TRUE ~ "ns")) 

cols <- c("up" = "gold", "down" = "blue", "ns" = "darkgrey") 
full_results_WS <- full_results_WS %>%
  mutate(gene_type = fct_relevel(gene_type, "up", "down")) 


## WR - Time T0 vs T1
full_results_WR <- full_results_WR %>%
  mutate(gene_type = case_when(logFC >= fc_cutoff & adj.P.Val <= 0.05 ~ "up",
                               logFC <= -fc_cutoff & adj.P.Val <= 0.05 ~ "down",
                               TRUE ~ "ns")) 

cols <- c("up" = "gold", "down" = "blue", "ns" = "darkgrey") 
full_results_WR <- full_results_WR %>%
  mutate(gene_type = fct_relevel(gene_type, "up", "down")) 


## NW - Time T0 vs T1
full_results_NW <- full_results_NW %>%
  mutate(gene_type = case_when(logFC >= fc_cutoff & adj.P.Val <= 0.05 ~ "up",
                               logFC <= -fc_cutoff & adj.P.Val <= 0.05 ~ "down",
                               TRUE ~ "ns")) 

cols <- c("up" = "gold", "down" = "blue", "ns" = "darkgrey") 
full_results_NW <- full_results_NW %>%
  mutate(gene_type = fct_relevel(gene_type, "up", "down")) 
```

## WS - Time T0 vs T1
```{r}
full_results_WS %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type)) + geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, " - Volcano plot - Comparaison WS - Time T0 vs T1"))
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-WS-Time.pdf'))

```

## WR - Time T0 vs T1
```{r}
full_results_WR %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type)) + geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, " - Volcano plot - Comparaison WR - Time T0 vs T1"))
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-WR-Time.pdf'))

```

## NW - Time T0 vs T1
```{r}
full_results_NW %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type)) + geom_point() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, " - Volcano plot - Comparaison NW - Time T0 vs T1"))
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-NW-Time.pdf'))

```


## Affichage des 20 gènes ayant leur expression la plus modifiée
```{r}
topN <- 20 #Nombre de genes affichés sur le graphe
```


```{r}
#WS - Time T0 vs T1
full_results_WS %>% 
  dplyr::mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type,label=Label)) + 
  geom_point() +    
  geom_text_repel(col = "black", max.overlaps = 25)  + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, ' - Volcano plot - WS - Time T0 vs T1 - Top ',topN,' genes '))
ggsave(paste0(my_id,'-Volcano-Comparaison-WS-Time-Top20-plot.pdf'))

filter(full_results_WS, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_WS-TIME_filtered_significant_results.csv"))
```


```{r}
#WS - Time T0 vs T1
full_results_WR %>% 
  dplyr::mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type,label=Label)) + 
  geom_point() +    
  geom_text_repel(col = "black", max.overlaps = 25)  + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, ' - Volcano plot - WR - Time T0 vs T1 - Top ',topN,' genes '))
ggsave(paste0(my_id,'-Volcano-Comparaison-WR-Time-Top20-plot.pdf'))

filter(full_results_WR, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_WR-TIME_filtered_significant_results.csv"))
```


```{r}
#NW - Time T0 vs T1
full_results_NW %>% 
  dplyr::mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=gene_type,label=Label)) + 
  geom_point() +    
  geom_text_repel(col = "black", max.overlaps = 25)  + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey"))+
  labs(title=paste0(my_id, ' - Volcano plot - NW - Time T0 vs T1 - Top ',topN,' genes '))
ggsave(paste0(my_id,'-Volcano-Comparaison-NW-Time-Top20-plot.pdf'))

filter(full_results_NW, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_NW-TIME_filtered_significant_results.csv"))
```



## Affichage des 20 gènes de la tolérance (Baron et al. 2015)
```{r}
my_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")
```

```{r}
#WS - Time T0 vs T1
full_results_WS %>% 
  dplyr::mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>%
  dplyr::mutate(Rank = rank(dplyr::desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = gene_type, label = Label)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey")) + 
  geom_text_repel(col = "black", max.overlaps = 5000) +
  labs(title="Volcano plot - WS - Time T0 vs T1 - 20 genes of tolerance (Baron et al. 2015)")
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-WS-Time-20-genes-TOL.pdf'))

filter(full_results_WS, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_WS-TIME_filtered_significant_results.csv"))
```


```{r}
#WR - Time T0 vs T1
full_results_WR %>% 
  dplyr::mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>%
  dplyr::mutate(Rank = rank(dplyr::desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = gene_type, label = Label)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey")) + 
  geom_text_repel(col = "black", max.overlaps = 5000) +
  labs(title="Volcano plot - WR - Time T0 vs T1 - 20 genes of tolerance (Baron et al. 2015)")
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-WR-Time-20-genes-TOL.pdf'))

filter(full_results_WR, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_WR-TIME_filtered_significant_results.csv"))
```


```{r}
#NW - Time T0 vs T1
full_results_NW %>% 
  dplyr::mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>%
  dplyr::mutate(Rank = rank(dplyr::desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = gene_type, label = Label)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") + 
  geom_vline(xintercept = c(-fc_cutoff, fc_cutoff),
             linetype = "dashed")+
  scale_color_manual(values=c("gold", "blue", "grey")) + 
  geom_text_repel(col = "black", max.overlaps = 5000) +
  labs(title="Volcano plot - NW - Time T0 vs T1 - 20 genes of tolerance (Baron et al. 2015)")
ggsave(paste0(my_id,'-Volcano-plot-Comparaison-NW-Time-20-genes-TOL.pdf'))

filter(full_results_NW, adj.P.Val < 0.05, abs(logFC) > fc_cutoff) %>%
  write_csv(path=paste0(my_id,"_NW-TIME_filtered_significant_results.csv"))
```







# Gene Ontology Term Enrichment analysis
```{r}
############# GO analysis ###########################
library("clusterProfiler")
library(org.Hs.eg.db)
library(DOSE)
# on rapelle que p_cutoff <- 0.05 
# et fc_cutoff <- 1 


DiffGenes <- full_results[ full_results$adj.P.Val  < p_cutoff & abs(full_results$logFC) > fc_cutoff, ]
dim(DiffGenes)
UpGenes <- full_results[ full_results$adj.P.Val  < p_cutoff & full_results$logFC > fc_cutoff, ]
dim(UpGenes)
DownGenes <- full_results[ full_results$adj.P.Val  < p_cutoff & full_results$logFC < -fc_cutoff, ]
dim(DownGenes)

DiffGenesSymb <- DiffGenes$GENE_SYMBOL
length(DiffGenesSymb)
UpGenesSymb <- UpGenes$GENE_SYMBOL
DownGenesSymb <- DownGenes$GENE_SYMBOL
geneList <- features$GENE_SYMBOL
length(geneList)

keytypes(org.Hs.eg.db)


ego <- enrichGO(gene         = UpGenesSymb,
                universe = geneList,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                #qvalueCutoff  = 0.5,
                minGSSize=5)

head(ego)
dim(ego)
data.frame(ego)


SelGO <- ego[which(ego$p.adjust  < 0.05 & ego$Count > 5),]

barplot(ego)
barplot(ego, showCategory= dim(SelGO)[1] )
barplot(ego, showCategory= 10)

write.table(data.frame(ego), file =paste("c-GO-", comparison, "-Down_diff-",".txt", sep=""), sep="\t",  na="NA", dec=".")

ego2 <- clusterProfiler::simplify(ego, cutoff=0.8, by="p.adjust", select_fun=min)
barplot(ego2)

jpeg(paste("h-GO-UP", comparison, ".jpeg", sep=""), width=6, height=6, units="in", res=400, quality=100)
dev.off()
#plotGOgraph(ego)
upsetplot(ego)
#cnetplot(ego, showCategory = 3, categorySize="pvalue", foldChange= DiffGenes$logFC)
cnetplot(ego, showCategory = 5, categorySize = "geneNum", foldChange = DiffGenes$logFC,   fixed = TRUE)



heatplot(ego, showCategory = 25) #, foldChange = DiffGenes$logFC)
heatplot(ego,showCategory = 25, foldChange = DiffGenes$logFC)


a <- data.frame( DiffGenes$GENE_SYMBOL, DiffGenes$logFC)
colnames(a) <- c("Symbol","logFC")

a <- a[!duplicated(a$Symbol), ]
rownames(a) <- a$Symbol


a <- cbind(as.character(DiffGenes$GENE_SYMBOL, DiffGenes$logFC))
rownames(a) <- a[,1]
a <- a[,-1]

emapplot(ego)

cnetplot(ego, foldChange=geneList)
```
