---
title: "R Notebook"
output: html_notebook
---

```{r eval=false}
#Library####
library(affycoretools)
library(RColorBrewer)
library(oligo)
library(ggplot2)
library(readxl)
library(GEOquery)
library(readr)
library(dplyr)
library(pheatmap)
library(ggrepel)
library(FactoMineR)
library(limma)


library("annotate")
library("heatmap3")
library("Biobase")
library("tools")
library("AnnotationDbi")
library("annotate")
library("heatmap3")
library("ade4")
library("ALL")
library("gplots")
library("ALL")
library("ade4")
library('gcrma')
library("RVAideMemoire")
library("affy")
library("pd.huex.1.0.st.v2")

#library("mixOmics")

#library("hugene10sttranscriptcluster")
#library("huex10sttranscriptcluster.db")
library("hgu133plus2.db")
#library(hugene10stv1cdf)
library(hgu133a.db)
```


```{r eval=false}
#Data import#####
wd_a <- "/Users/olivier/Desktop/Projet/Test/"
knitr::opts_chunk$set(root.dir = wd_a)
 # Accessing Raw Data from GEO:
my_id = "GSE22224"
#filePaths = getGEOSuppFiles(my_id)
```


```{r eval=false}
### Import data ----

#Lecture du fichier Target
file_pheno_xlsx <-paste0(my_id,"_Target.xlsx")
file_pheno <- paste0(my_id,"_Target.txt")

data <- read_excel(paste0(wd_a,my_id,"/",file_pheno_xlsx))

write.table(data, file = file_pheno, sep = "\t", row.names = TRUE)


pd <- read.AnnotatedDataFrame(file = file_pheno, row.names = 1, header = TRUE, as.is = TRUE, sep = "\t", col.names = c("GSM", "Group", "Patient", "Treatment", "Group_short", "Description", "File"))

# Iterate through each column of pd
for (col in colnames(pd)) {
  # Remove double quotes from the column
  pd[[col]] <- gsub("\"", "", pd[[col]])
}

```

```{r}
#Changement de wd (besoin d'extraire le fichier RAW)
wd <- paste0(wd_a,my_id,"/",my_id,"_RAW")
df1 <- getGSEDataTables #??


knitr::opts_chunk$set(root.dir = wd)
write.table(matrix(dir()),file="a-list-files.txt", sep="\t")
```

```{r}
#Reads CEL files


# Lectures des fichiers .CEL
rawAffyData <- read.celfiles(filenames = pData(pd)$File, phenoData = pd)

# Vérification du bon import et première visualisation des données brutes
exprs(rawAffyData)[1:4,1:3]
boxplot(rawAffyData, which=c("all"))
hist(rawAffyData)
hist(rawAffyData, which=c("all"))
#pmSeq <- pmSequence(rawAffyData)
#pmSeq[1:5]
colnames(rawAffyData)


# Normalizes data ####
#Summarised per gene:
eset <- oligo::rma(rawAffyData) #Normalization entre les puces
#boxplot(eset)
#change directory: ####
```


```{r}
# annotate: #### Selon la puce, changer le package d'annotation

eset <-annotateEset(eset,hgu133plus2.db)

# Vérification de la bonne annotation
geneName = "GAPDH"
which(fData(eset)$SYMBOL == geneName )
#head(fData(eset))


# filtration:
eset_medians <- rowMedians(Biobase::exprs(eset))
hist_res <- hist(eset_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities")
median(eset_medians)
man_threshold =median(eset_medians)
abline(v = median(eset_medians), col = "grey", lwd = 2)
abline(v = man_threshold, col = "coral4", lwd = 2)

length(which((rowMedians(exprs(eset)) >= man_threshold) == "TRUE"))

#Filtration selon la médiane d'expression (enleve aussi les NA)
data_filtr  <- eset[rowMedians(exprs(eset)) >= man_threshold,]
colnames(data_filtr) <- pData(data_filtr)$GSM
dim(data_filtr)
dim(eset)

#Rassemble tout dans un dataset
toto <- cbind(fData(data_filtr), exprs(data_filtr))
dim(toto)

#Export des données d'expression
#write.table(toto, file="c-affy-Norm-Glomeruli-8April21.txt", sep="\t")


##enlever gene sans Symbol: #### -> Mauvaise annotation
length(which(is.na(fData(data_filtr)$SYMBOL) ==T ))
index <- which(is.na(fData(data_filtr)$SYMBOL) ==T )
sansNA <- data_filtr[-index, ]
length(which(is.na(fData(sansNA)$SYMBOL) ==T ))
dim(sansNA)
dim(eset)
dim(data_filtr)

sansNA_medians <- rowMedians(Biobase::exprs(sansNA))
hist_res <- hist(sansNA_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities")
##### END import #####

```
## Analyse
```{r}
summary(exprs(data_filtr))
boxplot(exprs(data_filtr),outline=FALSE)
```

# Inspection des variables
```{r}
sampleInfo <- pData(data_filtr)
colnames(sampleInfo)
rownames(sampleInfo) <- sampleInfo$GSM
## Utilisation des infos contenues dans source_name_ch1 et characteristics_ch1.1 
sampleInfo <- dplyr::select(sampleInfo, Patient,Treatment,Group_short,Group,Description,GSM) 

## Si jamais on a besoin de renommer nos variables (plus lisible)
#sampleInfo <- dplyr::rename(sampleInfo,group = source_name_ch1, patient=characteristics_ch1)
View(sampleInfo)
```


# Heatmap entre GSM
```{r}
#Calcul de la matrice de correlation d'expression des gènes pour chaque GSM (étude de correlation entre les samples)
corMatrix <- cor(exprs(data_filtr),use="c")

## Vérifier que les colonnes de sampleInfo et corMatrix soient identiques
rownames(sampleInfo)
colnames(corMatrix)

## Sinon, on peut forcer avec le code suivant
rownames(sampleInfo) <- colnames(corMatrix)

# Supprimer les colonnes non nécessaires de sampleInfo
annotations <- sampleInfo[, c("Patient", "Treatment")]
# Récupérer l'ordre des lignes en fonction de sampleInfo$GSM
order_rows <- order(sampleInfo$GSM)

# Réorganiser les lignes de la matrice de corrélation
corMatrix_ordered <- corMatrix[order_rows, order_rows]

# Créer le heatmap en utilisant pheatmap avec les lignes ordonnées
pheatmap(corMatrix_ordered, annotation_col = annotations)
```

# Export

```{r eval=false}
#full_output <- cbind(fData(data_filtr),exprs(data_filtr))
#write_csv(full_output, path="gse_full_output.csv")
```

# Définition de nos features (et modification des noms de features si besoin (selon version de puces))
```{r}
#Import de features à partir des données fData
features <- fData(data_filtr)
colnames(features)

## Pour garder que certaines features
features <- dplyr::select(features,"SYMBOL","GENENAME","ENTREZID")

#Renommer les features (pour visualisation ensuite, il faut enlever les espaces)
features <- features %>% 
  rename(GENE_SYMBOL = `SYMBOL`,
         GENE_TITLE = `GENENAME`)

#Affichage de features
View(features)

#Export des features
full_output <- cbind(features,exprs(data_filtr))
write_csv(full_output, path="gse_full_output.csv")
```


## Differential Expression

# Création de notre matrice design et constrast
```{r}
# Création des données de design 
design_data <- data.frame(
  Patient = rep(1:5, each = 2),  # Répéter chaque patient deux fois
  GSM =  rownames(pData(data_filtr)),
  Time = rep(c("T0", "T1"), times = 5)  # Alterner entre T0 et T1 pour chaque patient
)

## Attention, vérifier que les GSM sont dans le bon ordre !
View(design_data)


##Création de la matrice de design 
# ici on confronte les 5 patients à T0 qui vont changer de traitement au nouveau traitement à T1 (T0+3mois)
design <- model.matrix(~0+ Time+Patient, data = design_data) 

# Afficher la matrice de design
print(design)
```


# ACP
```{r}

## Attention à bien avoir transposé la matrice d'expression sinon PCA en fonction des individus et non de l'expression des genes

#PCA classique
pca <- prcomp(t(exprs(data_filtr)))

##Rajouter les informations des patients
cbind(sampleInfo, pca$x) %>% 
ggplot(aes(x = PC1, y=PC2, col=Group,label=paste("Patient", Patient))) + 
  geom_point() + 
  geom_text_repel()
```


```{r}
#PCA en utilisant FactoMineR

#Calcul de l'ACP avec FactomineR
pca_bis <- PCA(X = t(exprs(data_filtr)), ncp = 5 , graph= FALSE)
```






## Etablissement de la différential expression

# lmFit
```{r}
fit <- lmFit(exprs(data_filtr), design)

#Vérifier qu'il n'y a pas de pb (sur les premiers coefficients)
head(fit$coefficients)
```


#Définition de nos contrasts pour la differencial expression

```{r}

## Si on fait T0 contre T1
contrasts <- makeContrasts( Comparaison=TimeT0 - TimeT1, levels = design)

# On applique nos constrasts 
fit2 <- contrasts.fit(fit, contrasts)
fit2

```

# On utilise *empirical Bayes'* pour avoir les p-values de chaque gènes (première estimation?)
```{r}
fit2 <- eBayes(fit2)

# Résultats avec topTable
topTable(fit2)

### pour voir les second constratst (si ils existtent? chaque individu T0 et T1??)
## topTable(fit2, coef=2)

#decideTests donne cb de genes sont differentially expressed
decideTests(fit2)

table(decideTests(fit2))
```


# Visualisation des resultat de la differencial expression

```{r}
fit2$genes <- features
summary(topTable(fit2))
```

```{r}
full_results <- topTable(fit2, number=Inf)
full_results <- tibble::rownames_to_column(full_results,"ID")
View(full_results)
```

```{r}
ggplot(full_results,aes(x = logFC, y= -log10(adj.P.Val))) + geom_point()
```


```{r}
p_cutoff <- 0.05 #CHANGER LE CUTOFF DE LA P-VALUE SI ON VEUX
fc_cutoff <- 1 #IDEM

#Problèmes car aucune significative???? (essais sans les poids???)

full_results %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant)) + geom_point()
```
## Affichage des 20 gènes ayant leur expression la plus modifiée

```{r}
p_cutoff <- 0.05
fc_cutoff <- 1
topN <- 20 #Nombre de genes affichés sur le graphe

full_results %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant,label=Label)) + geom_point() + geom_text_repel(col="black")
```
## Affichage des 20 gènes de la tolérance (Baron et al. 2015)
```{r}
my_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")

full_results %>% 
  mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>% 
  mutate(Rank = rank(desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = Significant, label = Label)) + 
  geom_point() + 
  geom_text_repel(col = "black", max.overlaps = 20)

#Problèmes --> non affichés??
```

## Filter selon pvalue ajustée?? (a faire)

```{r}
#p_cutoff <- 0.05
#fc_cutoff <- 1
#filter(full_results, adj.P.Val < 0.05, abs(logFC) > 1)
```



# Export avec write_csv (filtrer ceux significatifs)

```{r}
#filter(full_results, adj.P.Val < 0.05, abs(logFC) > 1) %>%
 # write_csv(path="filtered_de_results.csv")
```


### Autre visualisation (Heatmaps)


# Heatmap des 20 genes les plus significatifs
```{r}
## Top N genes

topN <- 20
##
top_ids_of_interest <- mutate(full_results, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(ID)

#On recupère le nom des top 20 genes
top_gene_names <- mutate(full_results, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(GENE_SYMBOL) 

## Recupere les expressions des genes d'interet
gene_matrix_top <- exprs(data_filtr)[top_ids_of_interest,]

##Affichage de la heatmpa des topN genes
pheatmap(gene_matrix_top,
     labels_row = top_gene_names,
     scale="row")

#PROBLEME MANQUE DEUX NOM DE GENES????
```


### Heatmap 20 genes de la tolérance (BARON)

```{r}
#20 GENES (BARON 2015)
Baron_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")

Baron_ids_of_interest <-  filter(full_results,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(ID)

Baron_gene_names <-  filter(full_results,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(GENE_SYMBOL)

Baron_gene_matrix <- exprs(data_filtr)[Baron_ids_of_interest,]

#Affichage
rownames(sampleInfo) <- colnames(Baron_gene_matrix)
pheatmap(Baron_gene_matrix,
         labels_row = Baron_gene_names,
         annotation_col= annotations,
         scale="row")
```


```{r}

#6 GENES (DANGER 2017)
Danger_genes <- c("AKR1C3", "CD40","CTLA4","ID3", "MZB1", "TCL1A")
#my_genes <- c("MS4A1", "BANK1")
#my_genes <- c("CD40","ID3", "MZB1", "TCL1A") #CTLA4/AKR1C3 marche pas??

Danger_ids_of_interest <-  filter(full_results,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(ID)

Danger_gene_names <-  filter(full_results,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(GENE_SYMBOL)

#Affichage
#rownames(sampleInfo) <- colnames(Danger_gene_matrix)
Danger_gene_matrix <- exprs(data_filtr)[Danger_ids_of_interest,]


pheatmap(Danger_gene_matrix,
         labels_row = Danger_gene_names,
         annotation_col= annotations,
         scale="row")
```
