---
title: "R Notebook"
output: html_notebook
---

```{r eval=false}
#Library####
library(affycoretools)
library(RColorBrewer)
library(oligo)
library(ggplot2)
library(readxl)
library(GEOquery)
library(readr)
library(dplyr)
library(pheatmap)
library(ggrepel)
#library(FactoMineR)
library(limma)


library("annotate")
library("heatmap3")
library("Biobase")
library("tools")
library("AnnotationDbi")
library("annotate")
library("heatmap3")
library("ade4")
library("ALL")
library("gplots")
library("ALL")
library("ade4")
library('gcrma')
library("RVAideMemoire")
library("affy")
library("pd.huex.1.0.st.v2")

#library("mixOmics")

#library("hugene10sttranscriptcluster")
#library("huex10sttranscriptcluster.db")
library("hgu133plus2.db")
#library(hugene10stv1cdf)
library(hgu133a.db)

# Fonction pour enregistrer les pheatmap (https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file)
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
    stopifnot(!missing(x))
    stopifnot(!missing(filename))
    pdf(filename, width=width, height=height)
    grid::grid.newpage()
    grid::grid.draw(x$gtable)
    dev.off()
}
```


```{r}
#Data import#####
wd_a <- "/Users/olivier/Desktop/Projet/Test/"
knitr::opts_chunk$set(root.dir = wd_a)
 # Accessing Raw Data from GEO:
my_id = "GSE22224"
#filePaths = getGEOSuppFiles(my_id)
```


```{r}
### Import data ----

#Lecture du fichier Target
file_pheno_xlsx <-paste0(my_id,"_Target.xlsx")
file_pheno <- paste0(my_id,"_Target.txt")

data <- read_excel(paste0(wd_a,my_id,"/",file_pheno_xlsx))

write.table(data, file = file_pheno, sep = "\t", row.names = TRUE)


pd <- read.AnnotatedDataFrame(file = file_pheno, row.names = 1, header = TRUE, as.is = TRUE, sep = "\t", col.names = c("GSM", "Group", "Patient", "Treatment", "Group_short", "Description", "File"))

# Iterate through each column of pd
for (col in colnames(pd)) {
  # Remove double quotes from the column
  pd[[col]] <- gsub("\"", "", pd[[col]])
}

```

```{r}
#Changement de wd (besoin d'extraire le fichier RAW)
wd <- paste0(wd_a,my_id,"/",my_id,"_RAW")
df1 <- getGSEDataTables #??


knitr::opts_chunk$set(root.dir = wd)
write.table(matrix(dir()),file="a-list-files.txt", sep="\t")
```

```{r}
#Reads CEL files


# Lectures des fichiers .CEL
rawAffyData <- read.celfiles(filenames = pData(pd)$File, phenoData = pd)

# Vérification du bon import et première visualisation des données brutes
exprs(rawAffyData)[1:4,1:3]
boxplot(rawAffyData, which=c("all"))
hist(rawAffyData)
hist(rawAffyData, which=c("all"))
#pmSeq <- pmSequence(rawAffyData)
#pmSeq[1:5]
colnames(rawAffyData)


# Normalizes data ####
#Summarised per gene:
eset <- oligo::rma(rawAffyData) #Normalization entre les puces
#boxplot(eset)
#change directory: ####
```


```{r}
# annotate: #### Selon la puce, changer le package d'annotation

eset <-annotateEset(eset,hgu133plus2.db)

# Vérification de la bonne annotation
geneName = "GAPDH"
which(fData(eset)$SYMBOL == geneName )
#head(fData(eset))


# filtration:
eset_medians <- rowMedians(Biobase::exprs(eset))
hist_res <- hist(eset_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities")
median(eset_medians)
man_threshold =median(eset_medians)
abline(v = median(eset_medians), col = "grey", lwd = 2)
abline(v = man_threshold, col = "coral4", lwd = 2)

length(which((rowMedians(exprs(eset)) >= man_threshold) == "TRUE"))

#Filtration selon la médiane d'expression (enleve aussi les NA)
data_filtr  <- eset[rowMedians(exprs(eset)) >= man_threshold,]
colnames(data_filtr) <- pData(data_filtr)$GSM
dim(data_filtr)
dim(eset)

#Rassemble tout dans un dataset
concat <- cbind(fData(data_filtr), exprs(data_filtr))
dim(concat)

#Export des données d'expression
#write.table(concat, file="c-affy-Norm-Glomeruli-8April21.txt", sep="\t")


##enlever gene sans Symbol: #### -> Mauvaise annotation
length(which(is.na(fData(data_filtr)$SYMBOL) ==T ))
index <- which(is.na(fData(data_filtr)$SYMBOL) ==T )
sansNA <- data_filtr[-index, ]
length(which(is.na(fData(sansNA)$SYMBOL) ==T ))
dim(sansNA)
dim(eset)
dim(data_filtr)

sansNA_medians <- rowMedians(Biobase::exprs(sansNA))
hist_res <- hist(sansNA_medians, 100, col = "cornsilk1", freq = FALSE, 
                 main = "Histogram of the median intensities", 
                 border = "antiquewhite4",
                 xlab = "Median intensities")
##### END import #####

```
## Analyse
```{r}
summary(exprs(data_filtr))
boxplot(exprs(data_filtr),outline=FALSE)
```

# Inspection des variables
```{r}
sampleInfo <- pData(data_filtr)
colnames(sampleInfo)
rownames(sampleInfo) <- sampleInfo$GSM
## Utilisation des infos contenues dans source_name_ch1 et characteristics_ch1.1 
sampleInfo <- dplyr::select(sampleInfo, Patient,Treatment,Group_short,Group,Description,GSM) 

## Si jamais on a besoin de renommer nos variables (plus lisible)
#sampleInfo <- dplyr::rename(sampleInfo,group = source_name_ch1, patient=characteristics_ch1)
View(sampleInfo)
```


# Heatmap entre GSM
```{r}
#Calcul de la matrice de correlation d'expression des gènes pour chaque GSM (étude de correlation entre les samples)
corMatrix <- cor(exprs(data_filtr),use="c")

## Vérifier que les colonnes de sampleInfo et corMatrix soient identiques
rownames(sampleInfo)
colnames(corMatrix)

## Sinon, on peut forcer avec le code suivant
rownames(sampleInfo) <- colnames(corMatrix)

# Supprimer les colonnes non nécessaires de sampleInfo
annotations <- sampleInfo[, c("Patient", "Treatment")]

# Récupérer l'ordre des lignes en fonction de sampleInfo$GSM
order_rows <- order(sampleInfo$GSM)

# Réorganiser les lignes de la matrice de corrélation
corMatrix_ordered <- corMatrix[order_rows, order_rows]

# Créer le heatmap en utilisant pheatmap avec les lignes ordonnées
pheatmap(corMatrix_ordered, annotation_col = annotations)
```

## Suppression des outliers
```{r}
#On remarque que le GSM553485 (HV), GSM553459 (CsA), GSM553458 (SRL) sont peu correlés aux restes des échantillons (probablement du à un effet batch, on décide de faire notre analse avec et sans ces samples.

GSM_outliers <- c("GSM553458")
data_outliers <- data_filtr[!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant
corMatrix_outliers <- cor(exprs(data_outliers), use = "c")
#corMatrix_outliers <- corMatrix_outliers [!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant

#Indices des colonnes à supprimer
colonnes_outliers <- which(colnames(corMatrix_outliers) %in% GSM_outliers)

# Indices des lignesà supprimer
lignes_outliers <- which(rownames(corMatrix_outliers) %in% GSM_outliers)

#Suppression des lignes et colonnes associés aux outliers
corMatrix_outliers <- corMatrix_outliers[-lignes_outliers, -colonnes_outliers]
pheatmap(corMatrix_outliers,
         annotation = annotations)


# Enlever HV et CsA
GSM_outliers <- c("GSM553485", "GSM553459")
data_outliers <- data_filtr[!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant
corMatrix_outliers <- cor(exprs(data_outliers), use = "c")
#corMatrix_outliers <- corMatrix_outliers [!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant

#Indices des colonnes à supprimer
colonnes_outliers <- which(colnames(corMatrix_outliers) %in% GSM_outliers)

# Indices des lignesà supprimer
lignes_outliers <- which(rownames(corMatrix_outliers) %in% GSM_outliers)

#Suppression des lignes et colonnes associés aux outliers
corMatrix_outliers <- corMatrix_outliers[-lignes_outliers, -colonnes_outliers]
pheatmap(corMatrix_outliers,
         annotation = annotations)

#Enlever HV et SRL
GSM_outliers <- c("GSM553485", "GSM553458")
data_outliers <- data_filtr[!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant
corMatrix_outliers <- cor(exprs(data_outliers), use = "c")
#corMatrix_outliers <- corMatrix_outliers [!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant

#Indices des colonnes à supprimer
colonnes_outliers <- which(colnames(corMatrix_outliers) %in% GSM_outliers)

# Indices des lignesà supprimer
lignes_outliers <- which(rownames(corMatrix_outliers) %in% GSM_outliers)

#Suppression des lignes et colonnes associés aux outliers
corMatrix_outliers <- corMatrix_outliers[-lignes_outliers, -colonnes_outliers]
pheatmap(corMatrix_outliers,
         annotation = annotations)


#Enlever les 3
GSM_outliers <- c("GSM553485", "GSM553459", "GSM553458")
data_outliers <- data_filtr[!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant
corMatrix_outliers <- cor(exprs(data_outliers), use = "c")
#corMatrix_outliers <- corMatrix_outliers [!data_filtr$GSM %in% GSM_outliers, ] #ne fonctionne pas pour l'instant

#Indices des colonnes à supprimer
colonnes_outliers <- which(colnames(corMatrix_outliers) %in% GSM_outliers)

# Indices des lignesà supprimer
lignes_outliers <- which(rownames(corMatrix_outliers) %in% GSM_outliers)

#Suppression des lignes et colonnes associés aux outliers
corMatrix_outliers <- corMatrix_outliers[-lignes_outliers, -colonnes_outliers]
pheatmap(corMatrix_outliers,
         annotation = annotations)
```

# Export
```{r eval=false}
#full_output <- cbind(fData(data_filtr),exprs(data_filtr))
#write_csv(full_output, path="gse_full_output.csv")
```

# Définition de nos features (et modification des noms de features si besoin (selon version de puces))
```{r}
#Import de features à partir des données fData
features <- fData(data_filtr)
colnames(features)

## Pour garder que certaines features
features <- dplyr::select(features,"SYMBOL","GENENAME","ENTREZID")

#Renommer les features (pour visualisation ensuite, il faut enlever les espaces)
features <- features %>% 
  rename(GENE_SYMBOL = `SYMBOL`,
         GENE_TITLE = `GENENAME`)

#Affichage de features
View(features)

#Export des features
full_output <- cbind(features,exprs(data_filtr))
write_csv(full_output, path="gse_full_output.csv")
```


## Differential Expression

# Création de notre matrice design et constrast
```{r}
# Création des données de design 
design_data <- data.frame(
  Patient = c(1:46),  # Répéter chaque patient deux fois
  GSM =  rownames(pData(data_filtr)),
  Treatment = pData(data_filtr)$Treatment  # Selon le traitement
)

## Attention, vérifier que les GSM sont dans le bon ordre !
View(design_data)


##Création de la matrice de design 
# ici on confronte les 5 patients à T0 qui vont changer de traitement au nouveau traitement à T1 (T0+3mois)
#design <- model.matrix(~0+ Treatment+Patient, data = design_data) 
design <- model.matrix(~0+ Treatment, data = design_data) 
# Afficher la matrice de design
print(design)
```


# ACP
```{r}

## Attention à bien avoir transposé la matrice d'expression sinon PCA en fonction des individus et non de l'expression des genes

#PCA classique
pca <- prcomp(t(exprs(data_filtr)))

##Rajouter les informations des patients
cbind(sampleInfo, pca$x) %>% 
ggplot(aes(x = PC1, y=PC2, col=Group,label=paste("Patient", Patient))) + 
  geom_point() + 
  geom_text_repel()
```


```{r}
#PCA en utilisant FactoMineR

#Calcul de l'ACP avec FactomineR
#pca_bis <- PCA(X = t(exprs(data_filtr)), ncp = 5 , graph= FALSE)
```



## Etablissement de la différential expression

# lmFit
```{r}
fit <- lmFit(exprs(data_filtr), design)

#Vérifier qu'il n'y a pas de pb (sur les premiers coefficients)
head(fit$coefficients)
```


#Définition de nos contrasts pour la differencial expression

```{r}

## CsA vs HV vs SRL
contrasts <- makeContrasts(EffetCsA=   TreatmentCsA-TreatmentHV,
                           DiffTrtmt=   TreatmentCsA-TreatmentSRL, 
                           EffetSRL = TreatmentHV-TreatmentSRL,
                           levels=colnames(design))

# On applique nos constrasts 

fit2 <- contrasts.fit(fit, contrasts)
fit2

```

# On utilise *empirical Bayes'* pour avoir les p-values de chaque gènes (première estimation?)
```{r}
fit2 <- eBayes(fit2)

# Résultats avec topTable
topTable(fit2)
comparisonCSA = "EffetCsA"
topTable(fit2, coef=comparisonCSA, adjust.method="BH")
comparisonSRL = "EffetSRL"
topTable(fit2, coef=comparisonSRL, adjust.method="BH")
comparisonDIFF = "DiffTrtmt"
topTable(fit2, coef=comparisonDIFF, adjust.method="BH")


#decideTests donne cb de genes sont differentially expressed
decideTests(fit2)
table(decideTests(fit2))
```


# Visualisation des resultat de la differencial expression

```{r}
fit2$genes <- features
summary(topTable(fit2))
```

```{r}
#full_results <- topTable(fit2, number=Inf)
full_results_CSA <- topTable(fit2, coef=comparisonCSA, adjust.method="BH", number=Inf)
full_results_SRL <- topTable(fit2, coef=comparisonSRL, adjust.method="BH", number=Inf)
full_results_DIFF <- topTable(fit2, coef=comparisonDIFF, adjust.method="BH", number=Inf)

full_results_CSA <- tibble::rownames_to_column(full_results_CSA,"ID")
full_results_SRL <- tibble::rownames_to_column(full_results_SRL,"ID")
full_results_DIFF <- tibble::rownames_to_column(full_results_DIFF,"ID")
#full_results <- tibble::rownames_to_column(full_results,"ID")

View(full_results_CSA)
#View(full_results)
```

## Affichage des volcano plots
```{r}
p_cutoff <- 0.05 #CHANGER LE CUTOFF DE LA P-VALUE SI ON VEUX
fc_cutoff <- 1 #IDEM
```

## CSA vs HV
```{r}
full_results_CSA %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant)) + geom_point() + ggtitle("GSE22224 - Comparaison CSA vs HV")

```
## SRL vs HV
```{r}
full_results_SRL %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant)) + geom_point() + ggtitle("GSE22224 - Comparaison SRL vs HV")
```

## CSA vs SRL
```{r}
full_results_DIFF %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant)) + geom_point() + ggtitle("GSE22224 - Comparaison CSA vs HV")
```


## Affichage des 20 gènes ayant leur expression la plus modifiée
```{r}
topN <- 20 #Nombre de genes affichés sur le graphe
```


```{r}
#CSA vs HV
full_results_CSA %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant,label=Label)) + geom_point() + geom_text_repel(col="black")

filter(full_results_CSA, adj.P.Val < 0.05, abs(logFC) > 1) %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-HV_filtered_significant_results.csv"))

#SRL vs HV
full_results_SRL %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant,label=Label)) + geom_point() + geom_text_repel(col="black")

filter(full_results_SRL, adj.P.Val < 0.05, abs(logFC) > 1) %>%
  write_csv(path=paste0('GSE_',my_id,"_SRL-HV_filtered_significant_results.csv"))

#CSA VS SRL
full_results_DIFF %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, GENE_SYMBOL,"")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col=Significant,label=Label)) + geom_point() + geom_text_repel(col="black")

filter(full_results_DIFF, adj.P.Val < 0.05, abs(logFC) > 1) %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-SRL_filtered_significant_results.csv"))
```



## Affichage des 20 gènes de la tolérance (Baron et al. 2015)
```{r}
my_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")
```

```{r}
#CSA vs HV
full_results_CSA %>% 
  mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>% 
  mutate(Rank = rank(desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = Significant, label = Label)) + 
  geom_point() + 
  geom_text_repel(col = "black", max.overlaps = 20)

#SRL vs HV
full_results_SRL %>% 
  mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>% 
  mutate(Rank = rank(desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = Significant, label = Label)) + 
  geom_point() + 
  geom_text_repel(col = "black", max.overlaps = 20)

#CSA vs SRL
full_results_DIFF %>% 
  mutate(Significant = adj.P.Val < p_cutoff & abs(logFC) > fc_cutoff) %>% 
  mutate(Rank = rank(desc(abs(logFC))), Label = ifelse(GENE_SYMBOL %in% my_genes, as.character(GENE_SYMBOL), "")) %>% 
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), col = Significant, label = Label)) + 
  geom_point() + 
  geom_text_repel(col = "black", max.overlaps = 20)
```

### Autre visualisation (Heatmaps)


# Heatmap des 20 genes les plus significatifs
```{r}
## Top N genes
topN <- 20


## CSA vs HV
top_ids_of_interest_CSA <- mutate(full_results_CSA, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(ID)

#On recupère le nom des top 20 genes
top_gene_names_CSA <- mutate(full_results_CSA, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(GENE_SYMBOL) 

## Recupere les expressions des genes d'interet
gene_matrix_top_CSA <- exprs(data_filtr)[top_ids_of_interest_CSA,] %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-HV_filtered_results_topN_genes.csv"))

##Affichage de la heatmpa des topN genes
CSA_HV_heatmap_topN_genes <- pheatmap(gene_matrix_top_CSA,
     labels_row = top_gene_names_CSA,
     annotation_col= annotations,
     scale="row")

save_pheatmap_pdf(CSA_HV_heatmap_topN_genes, paste0('GSE_',my_id,'_CSA-HV_heatmap_topN_genes.pdf'))



## SRL vs HV
top_ids_of_interest_SRL <- mutate(full_results_SRL, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(ID)

#On recupère le nom des top 20 genes
top_gene_names_SRL <- mutate(full_results_SRL, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(GENE_SYMBOL) 

## Recupere les expressions des genes d'interet
gene_matrix_top_SRL <- exprs(data_filtr)[top_ids_of_interest_SRL,] %>%
  write_csv(path=paste0('GSE_',my_id,"_SRL-HV_filtered_results_topN_genes.csv"))

##Affichage de la heatmpa des topN genes
SRL_HV_heatmap_topN_genes <- pheatmap(gene_matrix_top_SRL,
     labels_row = top_gene_names_SRL,
     annotation_col= annotations,
     scale="row")

save_pheatmap_pdf(SRL_HV_heatmap_topN_genes, paste0('GSE_',my_id,'_SRL-HV_heatmap_topN_genes.pdf'))


## CSA vs SRL
top_ids_of_interest_DIFF <- mutate(full_results_DIFF, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(ID)

#On recupère le nom des top 20 genes
top_gene_names_DIFF <- mutate(full_results_DIFF, Rank = 1:n()) %>% 
  filter(Rank < topN) %>% 
  pull(GENE_SYMBOL) 

## Recupere les expressions des genes d'interet
gene_matrix_top_DIFF <- exprs(data_filtr)[top_ids_of_interest_DIFF,] %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-SRL_filtered_results_topN_genes.csv")) 

##Affichage de la heatmpa des topN genes
CSA_SRL_heatmap_topN_genes <- pheatmap(gene_matrix_top_DIFF,
     labels_row = top_gene_names_DIFF,
     annotation_col= annotations,
     scale="row") 

save_pheatmap_pdf(CSA_SRL_heatmap_topN_genes, paste0('GSE_',my_id,'_CSA-SRL_heatmap_topN_genes.pdf'))
```


### Heatmap 20 genes de la tolérance (BARON)

```{r}
#20 GENES (BARON 2015)
Baron_genes <- c("TCL1A", "MZB1", "CD22", "BLK", "MS4A1", "CD79B", "BLNK", "FCRL2", "IRF4", "ID3", "AKR1C3", "HINT1", "ANXA2R","CD40", "FCER2", "CTLA4", "AKIRIN2", "EPS15", "PLBD1")



# CSA vs HV
Baron_ids_of_interest_CSA <-  filter(full_results_CSA,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(ID)

Baron_gene_names_CSA <-  filter(full_results_CSA,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(GENE_SYMBOL)

Baron_gene_matrix_CSA <- exprs(data_filtr)[Baron_ids_of_interest_CSA,]  %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-HV_filtered_results_20_genes.csv"))

#Affichage
#rownames(sampleInfo) <- colnames(Baron_gene_matrix_CSA)
CSA_HV_heatmap_20_genes <- pheatmap(Baron_gene_matrix_CSA,
         labels_row = Baron_gene_names_CSA,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(CSA_HV_heatmap_20_genes, paste0('GSE_',my_id,'_CSA-HV_heatmap_20_genes.pdf'))



# SRL vs HV
Baron_ids_of_interest_SRL <-  filter(full_results_SRL,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(ID)

Baron_gene_names_SRL <-  filter(full_results_SRL,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(GENE_SYMBOL)

Baron_gene_matrix_SRL <- exprs(data_filtr)[Baron_ids_of_interest_SRL,] %>%
  write_csv(path=paste0('GSE_',my_id,"_SRL-HV_filtered_results_20_genes.csv"))

#Affichage
#rownames(sampleInfo) <- colnames(Baron_gene_matrix_SRL)
SRL_HV_heatmap_20_genes <- pheatmap(Baron_gene_matrix_SRL,
         labels_row = Baron_gene_names_SRL,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(SRL_HV_heatmap_20_genes, paste0('GSE_',my_id,'_SRL-HV_heatmap_20_genes.pdf'))



# CSA vs SRL
Baron_ids_of_interest_DIFF <-  filter(full_results_DIFF,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(ID)

Baron_gene_names_DIFF <-  filter(full_results_DIFF,GENE_SYMBOL %in% Baron_genes) %>% 
  pull(GENE_SYMBOL)

Baron_gene_matrix_DIFF <- exprs(data_filtr)[Baron_ids_of_interest_DIFF,]  %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-SRL_filtered_results_20_genes.csv"))

#Affichage
#rownames(sampleInfo) <- colnames(Baron_gene_matrix__DIFF)
CSA_SRL_heatmap_20_genes <- pheatmap(Baron_gene_matrix_DIFF,
         labels_row = Baron_gene_names_DIFF,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(CSA_SRL_heatmap_20_genes, paste0('GSE_',my_id,'_CSA-SRL_heatmap_20_genes.pdf'))


```


```{r}

#6 GENES (DANGER 2017)
Danger_genes <- c("AKR1C3", "CD40","CTLA4","ID3", "MZB1", "TCL1A")
#my_genes <- c("MS4A1", "BANK1")
#my_genes <- c("CD40","ID3", "MZB1", "TCL1A") #CTLA4/AKR1C3 marche pas??



# CSA vs HV
Danger_ids_of_interest_CSA <-  filter(full_results_CSA,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(ID)

Danger_gene_names_CSA <-  filter(full_results_CSA,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(GENE_SYMBOL)

#Affichage
#rownames(sampleInfo) <- colnames(Danger_gene_matrix)
Danger_gene_matrix_CSA <- exprs(data_filtr)[Danger_ids_of_interest_CSA,] %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-HV_filtered_results_6_genes.csv"))


heatmap_6_genes_CSA <- pheatmap(Danger_gene_matrix_CSA,
         labels_row = Danger_gene_names_CSA,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(heatmap_6_genes_CSA, paste0('GSE_',my_id,'_CSA-HV_heatmap_6_genes.pdf'))


# CSA vs HV
Danger_ids_of_interest_SRL <-  filter(full_results_SRL,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(ID)

Danger_gene_names_SRL <-  filter(full_results_SRL,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(GENE_SYMBOL)

#Affichage
#rownames(sampleInfo) <- colnames(Danger_gene_matrix)
Danger_gene_matrix_SRL <- exprs(data_filtr)[Danger_ids_of_interest_SRL,] %>%
  write_csv(path=paste0('GSE_',my_id,"_SRL-HV_filtered_results_6_genes.csv"))


heatmap_6_genes_SRL <- pheatmap(Danger_gene_matrix_SRL,
         labels_row = Danger_gene_names_SRL,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(heatmap_6_genes_SRL, paste0('GSE_',my_id,'_SRL-HV_heatmap_6_genes.pdf'))



# CSA vs HV
Danger_ids_of_interest_DIFF <-  filter(full_results_DIFF,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(ID)

Danger_gene_names_DIFF <-  filter(full_results_DIFF,GENE_SYMBOL %in% Danger_genes) %>% 
  pull(GENE_SYMBOL)

#Affichage
#rownames(sampleInfo) <- colnames(Danger_gene_matrix)
Danger_gene_matrix_DIFF <- exprs(data_filtr)[Danger_ids_of_interest_DIFF,] %>%
  write_csv(path=paste0('GSE_',my_id,"_CSA-SRL_filtered_results_6_genes.csv"))


heatmap_6_genes_DIFF <- pheatmap(Danger_gene_matrix_DIFF,
         labels_row = Danger_gene_names_DIFF,
         annotation_col= annotations,
         scale="row")

save_pheatmap_pdf(heatmap_6_genes_DIFF, paste0('GSE_',my_id,'_CSA-SRL_heatmap_6_genes.pdf'))
```

# Gene Ontology Term Enrichment analysis
```{r}
############# GO analysis ###########################
library("clusterProfiler")
library(org.Hs.eg.db)
library(DOSE)
# on rapelle que p_cutoff <- 0.05 
# et fc_cutoff <- 1 


DiffGenes_CSA <- full_results_CSA[ full_results_CSA$adj.P.Val  < p_cutoff & abs(full_results_CSA$logFC) > fc_cutoff, ]
dim(DiffGenes_CSA)
UpGenes_CSA <- full_results_CSA[ full_results_CSA$adj.P.Val  < p_cutoff & full_results_CSA$logFC > fc_cutoff, ]
dim(UpGenes_CSA)
DownGenes_CSA <- full_results_CSA[ full_results_CSA$adj.P.Val  < p_cutoff & full_results_CSA$logFC < -fc_cutoff, ]
dim(DownGenes_CSA)

DiffGenesSymb_CSA <- DiffGenes_CSA$GENE_SYMBOL
length(DiffGenesSymb_CSA)
UpGenesSymb_CSA <- UpGenes_CSA$GENE_SYMBOL
DownGenesSymb_CSA <- DownGenes_CSA$GENE_SYMBOL
geneList <- features$GENE_SYMBOL
length(geneList)

keytypes(org.Hs.eg.db)


ego_CSA <- enrichGO(gene         = UpGenesSymb_CSA,
                universe = geneList,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                #qvalueCutoff  = 0.5,
                minGSSize=5)

head(ego_CSA)
dim(ego_CSA)
data.frame(ego_CSA)


SelGO_CSA <- ego_CSA[which(ego_CSA$p.adjust  < 0.05 & ego_CSA$Count > 5),]

barplot(ego_CSA)
barplot(ego_CSA, showCategory= dim(SelGO_CSA)[1] )
barplot(ego_CSA, showCategory= 10)

write.table(data.frame(ego_CSA), file =paste("c-GO-", comparisonCSA, "-Down_diff-",".txt", sep=""), sep="\t",  na="NA", dec=".")

ego2_CSA <- clusterProfiler::simplify(ego_CSA, cutoff=0.8, by="p.adjust", select_fun=min)
barplot(ego2_CSA)

jpeg(paste("h-GO-UP", comparisonCSA, ".jpeg", sep=""), width=6, height=6, units="in", res=400, quality=100)
dev.off()
#plotGOgraph(ego)
upsetplot(ego_CSA)
#cnetplot(ego, showCategory = 3, categorySize="pvalue", foldChange= DiffGenes$logFC)
cnetplot(ego_CSA, showCategory = 5, categorySize = "geneNum", foldChange = DiffGenes_CSA$logFC,   fixed = TRUE)



heatplot(ego_CSA, showCategory = 25) #, foldChange = DiffGenes$logFC)
heatplot(ego_CSA,showCategory = 25, foldChange = DiffGenes_CSA$logFC)


a_CSA <- data.frame( DiffGenes_CSA$GENE_SYMBOL, DiffGenes_CSA$logFC)
colnames(a_CSA) <- c("Symbol","logFC")

a_CSA <- a_CSA[!duplicated(a_CSA$Symbol), ]
rownames(a_CSA) <- a_CSA$Symbol


a_CSA <- cbind(as.character(DiffGenes_CSA$GENE_SYMBOL, DiffGenes_CSA$logFC))
rownames(a_CSA) <- a_CSA[,1]
a_CSA <- a_CSA[,-1]

emapplot(ego_CSA)

cnetplot(ego_CSA, foldChange=geneList)
```


```{r}

DiffGenes_SRL <- full_results_SRL[ full_results_SRL$adj.P.Val  < p_cutoff & abs(full_results_SRL$logFC) > fc_cutoff, ]
dim(DiffGenes_SRL)
UpGenes_SRL <- full_results_SRL[ full_results_SRL$adj.P.Val  < p_cutoff & full_results_SRL$logFC > fc_cutoff, ]
dim(UpGenes_SRL)
DownGenes_SRL <- full_results_SRL[ full_results_SRL$adj.P.Val  < p_cutoff & full_results_SRL$logFC < -fc_cutoff, ]
dim(DownGenes_SRL)

DiffGenesSymb_SRL <- DiffGenes_SRL$GENE_SYMBOL
length(DiffGenesSymb_SRL)
UpGenesSymb_SRL <- UpGenes_SRL$GENE_SYMBOL
DownGenesSymb_SRL <- DownGenes_SRL$GENE_SYMBOL
geneList <- features$GENE_SYMBOL
length(geneList)

keytypes(org.Hs.eg.db)


ego_SRL <- enrichGO(gene         = UpGenesSymb_SRL,
                universe = geneList,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                #qvalueCutoff  = 0.5,
                minGSSize=5)

head(ego_SRL)
dim(ego_SRL)
data.frame(ego_SRL)


SelGO_SRL <- ego_SRL[which(ego_SRL$p.adjust  < 0.05 & ego_SRL$Count > 5),]

barplot(ego_SRL)
barplot(ego_SRL, showCategory= dim(SelGO_SRL)[1] )
barplot(ego_SRL, showCategory= 10)

write.table(data.frame(ego_SRL), file =paste("c-GO-", comparisonCSA, "-Down_diff-",".txt", sep=""), sep="\t",  na="NA", dec=".")

ego2_SRL <- clusterProfiler::simplify(ego_SRL, cutoff=0.8, by="p.adjust", select_fun=min)
barplot(ego2_SRL)

jpeg(paste("h-GO-UP", comparisonCSA, ".jpeg", sep=""), width=6, height=6, units="in", res=400, quality=100)
dev.off()
#plotGOgraph(ego)
upsetplot(ego_SRL)
#cnetplot(ego, showCategory = 3, categorySize="pvalue", foldChange= DiffGenes$logFC)
cnetplot(ego_SRL, showCategory = 5, categorySize = "geneNum", foldChange = DiffGenes_SRL$logFC,   fixed = TRUE)



heatplot(ego_SRL, showCategory = 25) #, foldChange = DiffGenes$logFC)
heatplot(ego_SRL,showCategory = 25, foldChange = DiffGenes_SRL$logFC)


a_SRL <- data.frame( DiffGenes_SRL$GENE_SYMBOL, DiffGenes_SRL$logFC)
colnames(a_SRL) <- c("Symbol","logFC")

a_SRL <- a_SRL[!duplicated(a_SRL$Symbol), ]
rownames(a_SRL) <- a_SRL$Symbol


a_SRL <- cbind(as.character(DiffGenes_SRL$GENE_SYMBOL, DiffGenes_SRL$logFC))
rownames(a_SRL) <- a_SRL[,1]
a_SRL <- a_SRL[,-1]

emapplot(ego_SRL)

cnetplot(ego_SRL, foldChange=geneList)
```


```{r}

DiffGenes_DIFF <- full_results_DIFF[ full_results_DIFF$adj.P.Val  < p_cutoff & abs(full_results_DIFF$logFC) > fc_cutoff, ]
dim(DiffGenes_DIFF)
UpGenes_DIFF <- full_results_DIFF[ full_results_DIFF$adj.P.Val  < p_cutoff & full_results_DIFF$logFC > fc_cutoff, ]
dim(UpGenes_DIFF)
DownGenes_DIFF <- full_results_DIFF[ full_results_DIFF$adj.P.Val  < p_cutoff & full_results_DIFF$logFC < -fc_cutoff, ]
dim(DownGenes_DIFF)

DiffGenesSymb_DIFF <- DiffGenes_DIFF$GENE_SYMBOL
length(DiffGenesSymb_DIFF)
UpGenesSymb_DIFF <- UpGenes_DIFF$GENE_SYMBOL
DownGenesSymb_DIFF <- DownGenes_DIFF$GENE_SYMBOL
geneList <- features$GENE_SYMBOL
length(geneList)

keytypes(org.Hs.eg.db)


ego_DIFF <- enrichGO(gene         = UpGenesSymb_DIFF,
                universe = geneList,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                #qvalueCutoff  = 0.5,
                minGSSize=5)

head(ego_DIFF)
dim(ego_DIFF)
data.frame(ego_DIFF)


SelGO_DIFF <- ego_DIFF[which(ego_DIFF$p.adjust  < 0.05 & ego_DIFF$Count > 5),]

barplot(ego_DIFF)
barplot(ego_DIFF, showCategory= dim(SelGO_DIFF)[1] )
barplot(ego_DIFF, showCategory= 10)

write.table(data.frame(ego_DIFF), file =paste("c-GO-", comparisonCSA, "-Down_diff-",".txt", sep=""), sep="\t",  na="NA", dec=".")

ego2_DIFF <- clusterProfiler::simplify(ego_DIFF, cutoff=0.8, by="p.adjust", select_fun=min)
barplot(ego2_DIFF)

jpeg(paste("h-GO-UP", comparisonCSA, ".jpeg", sep=""), width=6, height=6, units="in", res=400, quality=100)
dev.off()
#plotGOgraph(ego)
upsetplot(ego_DIFF)
#cnetplot(ego, showCategory = 3, categorySize="pvalue", foldChange= DiffGenes$logFC)
cnetplot(ego_DIFF, showCategory = 5, categorySize = "geneNum", foldChange = DiffGenes_DIFF$logFC,   fixed = TRUE)



heatplot(ego_DIFF, showCategory = 25) #, foldChange = DiffGenes$logFC)
heatplot(ego_DIFF,showCategory = 25, foldChange = DiffGenes_DIFF$logFC)


a_DIFF <- data.frame( DiffGenes_DIFF$GENE_SYMBOL, DiffGenes_DIFF$logFC)
colnames(a_DIFF) <- c("Symbol","logFC")

a_DIFF <- a_DIFF[!duplicated(a_DIFF$Symbol), ]
rownames(a_DIFF) <- a_DIFF$Symbol


a_DIFF <- cbind(as.character(DiffGenes_DIFF$GENE_SYMBOL, DiffGenes_DIFF$logFC))
rownames(a_DIFF) <- a_DIFF[,1]
a_DIFF <- a_DIFF[,-1]

emapplot(ego_DIFF)

cnetplot(ego_DIFF, foldChange=geneList)
```

